psl-core/src/main/java/org/linqs/psl/model/rule/AbstractRule.java

```diff
@@ -23,45 +23,49 @@
 
 /**
  * Base class for all (first order, i.e., not ground) rules.
- *
- * Care should always be taken when creating multiple instance of the same rule (e.g. through deserialization) or checking rules for equality.
- * In general, there should only be one instance of each rule, but this is not tightly enforced.
- * Rule hashes are by default their identity hash code (object id).
- * This means that it may be possible to have rules that are equal(), but that do not have matching hashes.
  */
 public abstract class AbstractRule implements Rule {
-    private static Map<Integer, AbstractRule> rules = new HashMap<Integer, AbstractRule>();
+    private static final Map<Integer, Rule> rules = new HashMap<Integer, Rule>();
 
     protected final String name;
+    private int hashcode;
 
-    public static AbstractRule getRule(int hashcode) {
+    public static Rule getRule(int hashcode) {
         return rules.get(hashcode);
     }
 
-    public AbstractRule(String name) {
-        this(name, true);
-    }
-
-    protected AbstractRule(String name, boolean storeRule) {
+    protected AbstractRule(String name, int hashcode) {
         this.name = name;
+        this.hashcode = hashcode;
+
+        rules.put(hashcode, this);
+    }
 
-        if (storeRule) {
-            rules.put(System.identityHashCode(this), this);
-        }
+    protected void setHashcode(int hashcode) {
+        rules.remove(this.hashcode);
+        this.hashcode = hashcode;
+        rules.put(hashcode, this);
     }
 
     public String getName() {
         return this.name;
     }
 
-    /**
-     * Explicitly define the hash for rules to be the identity hash code (object id).
-     */
+    public static void registerRule(Rule rule) {
+        rules.put(rule.hashCode(), rule);
+    }
+
+    public static void deleteRule(Rule rule) {
+        rules.remove(rule.hashCode());
+    }
+
```

Comment:
If we can, these should both be non-static.
The rule already exists, so no need to pass it.